Refine the API in the public header

ChinYikMing opened the pull request:
The following should be included in an emulator's simple and clear public API:
1. create/init core
2. run emulation
3. delete/destroy core

Other components, including as memory, file systems, program data, etc., should be abstracted from the user, as a result, setting a configuration value (`vm_attr_t`) is sufficient. The user should manage about memory (`state_t`) and elf stuff before this PR. The user may just construct a core, run it, and shut it down after this PR, so they won't need to worry about them anymore.

For stdio remapping, `rv_remap_stdstream` function is introduced.

The `vm_attr_t` has multiple fields and they are commented clearly in the code.

elf is reopened in `dump_test_signature` because elf is allocated during `rv_create`. It is acceptable to reopen elf since it is only for testing. Print inferior exit code to console inside main instead of `syscall_exit` because the actual usage of exit code depends on applications of using riscv public API.

The `io` interface is not changed in this PR because it could maybe reused with `semu` in some way, still need to be investigated. Also, Logging feature and system emulator integration are not implemented yet.

ChinYikMing commented on src/syscall_sdl.c:
```diff
@@ -149,7 +149,7 @@ static submission_queue_t submission_queue = {
 static submission_t submission_pop(riscv_t *rv)
 {
-    state_t *s = rv_userdata(rv);
+    vm_attr_t *s = rv_userdata(rv);
```
`s` should be renamed to `vm_attr`

ChinYikMing commented on src/syscall_sdl.c:
```diff
@@ -162,7 +162,7 @@ static submission_t submission_pop(riscv_t *rv)
 static void event_push(riscv_t *rv, event_t event)
 {
-    state_t *s = rv_userdata(rv);
+    vm_attr_t *s = rv_userdata(rv);
```
`s` should be renamed to `vm_attr`

ChinYikMing commented on src/syscall_sdl.c:
```diff
@@ -279,7 +279,7 @@ static bool check_sdl(riscv_t *rv, int width, int height)
 void syscall_draw_frame(riscv_t *rv)
 {
-    state_t *s = rv_userdata(rv); /* access userdata */
+    vm_attr_t *s = rv_userdata(rv); /* access userdata */
```
`s` should be renamed to `vm_attr`

ChinYikMing commented on src/syscall_sdl.c:
```diff
@@ -677,7 +677,7 @@ static void *music_handler(void *arg)
 static void play_sfx(riscv_t *rv)
 {
-    state_t *s = rv_userdata(rv); /* access userdata */
+    vm_attr_t *s = rv_userdata(rv); /* access userdata */
```
`s` should be renamed to `vm_attr`

ChinYikMing commented on src/syscall_sdl.c:
```diff
@@ -705,7 +705,7 @@ static void play_sfx(riscv_t *rv)
 static void play_music(riscv_t *rv)
 {
-    state_t *s = rv_userdata(rv); /* access userdata */
+    vm_attr_t *s = rv_userdata(rv); /* access userdata */
```
`s` should be renamed to `vm_attr`

jserv replied:
Or, `attr`.

jserv commented on src/emulate.c:
```diff
@@ -1182,15 +1183,15 @@ void ecall_handler(riscv_t *rv)
 void memset_handler(riscv_t *rv)
 {
-    memory_t *m = ((state_t *) rv->userdata)->mem;
+    memory_t *m = ((vm_attr_t *) rv->userdata)->mem;
```
We can introduce a helper macro similar to how `PRIV` is used in semu. That is,
```c
#define PRIV(x) ((vm_attr_t *) x->userdata)
```

jserv commented on src/main.c:
```diff
@@ -221,12 +215,19 @@ int main(int argc, char **args)
-        fprintf(stderr, "Unable to open ELF file '%s'\n", opt_prog_name);
-        return 1;
-    }
+    vm_attr_t vm_attr = {
```
Name the instance of `vm_attr_t` simply as `attr`.

jserv commented on src/riscv.c:
```diff
@@ -105,11 +106,57 @@ riscv_word_t rv_get_reg(riscv_t *rv, uint32_t reg)
+ * @new_fptr: FILE * for new redirection
+ *
+ */
+void rv_register_stdio(riscv_t *rv, int stdio_fd, FILE *new_fptr)
```
Quote [Wikipedia](https://en.wikipedia.org/wiki/Standard_streams):
> In computer programming, standard streams are preconnected input and output communication channels between a computer program and its environment when it begins execution. The three input/output (I/O) connections are called standard input (stdin), standard output (stdout) and standard error (stderr).

In the proposed changed code, the use of stdio is confusing.

ChinYikMing replied:
I think `rv_remap_stream(riscv_t *rv, int fd, FILE *fptr)` is more comprehensive. More comments can be added to classify the parameters.

jserv replied:
How about `rv_remap_stdstream(riscv_t *rv, int fd, FILE *filev)`?
Here `filev` means a vector of `FILE *`.

ChinYikMing replied:
A fd is given in a call, is it necessary required a vector of `FILE *`?

jserv replied:
> A fd is given in a call, is it necessary required a vector of `FILE *`?

No, my intention was to encapsulate the pair of a file descriptor and standard streams. This was meant to form a vector, but I misplaced the parameters.

ChinYikMing replied:
> How about rv_remap_stdstream(riscv_t *rv, int fd, FILE *filev)?

Much more sense than `rv_register_stdio`.

> > A fd is given in a call, is it necessary required a vector of `FILE *`?
> 
> No, my intention was to encapsulate the pair of a file descriptor and standard streams. This was meant to form a vector, but I misplaced the parameters.

Got it. So, the signature becomes `rv_remap_stdstream(riscv_t *rv, int *fdv, FILE *filev)`. We could assume the length of the vector is 3, otherwise undefined behavior. If any `fd` in `fdv` is -1, it fallback to use default stdio.

jserv replied:
> So, the signature becomes `rv_remap_stdstream(riscv_t *rv, int *fdv, FILE *filev)`. We could assume the length of the vector is 3, otherwise undefined behavior. If any `fd` in `fdv` is -1, it fallback to use default stdio.

To mitigate the risk of passing null or malformed pointers, we could introduce a structure, `fd_stream_pair`, to encapsulate file descriptors and `FILE` pointers. This structure can then be reliably passed as a pointer to the `rv_remap_stdstream` function, ensuring safer and more organized handling of these elements.

ChinYikMing commented:
As discuss before, exit code shall be stored and maybe not to output to console directly. 

Unfortunately, CI/host-x64 and CI/host-arm run `make check` and `make misalign` for testing, and the compare signature includes "inferior exit code 0" during exit but I do not output them in this PR. Shall we modify `make check` and `make misalign` ?

jserv commented on src/syscall.c:
```diff
@@ -376,4 +380,12 @@ void syscall_handler(riscv_t *rv)
+     *
+     * the application decides the usage of the return code
+     */
+    vm_attr_t *vm_attr = rv_userdata(rv); /* access userdata */
```
The line `vm_attr_t *vm_attr = rv_userdata(rv);` might be perplexing due to the similar naming of the object type and the `vm_attr`, while the value comes from the `rv_userdata` call. This suggests a presumed direct link between `rv_userdata` and `vm_attr_t`. However, this connection is not explicitly clarified in the API documentation, leading to potential confusion regarding their relationship.

ChinYikMing replied:
Added more comments about them. Hopefully it is clearer now.

jserv commented:
> As discuss before, exit code shall be stored and maybe not to output to console directly.
> Unfortunately, CI/host-x64 and CI/host-arm run `make check` and `make misalign` for testing, and the compare signature includes "inferior exit code 0" during exit but I do not output them in this PR. Shall we modify `make check` and `make misalign` ?

After a RISC-V program ends, it is important to verify the exit code (i.e., `$?`). Consequently, modifications to 'check' and related build targets are necessary to accommodate this.

ChinYikMing commented:
> After a RISC-V program ends, it is important to verify the exit code (i.e., `$?`). Consequently, modifications to 'check' and related build targets are necessary to accommodate this.

Sure, a build macro can be added to `syscall_exit` and print  the exit code only on test for passing CI.

visitorckw commented:
Due to the memory size being changed to a runtime-defined value, do we need to check if the addresses of load and store instructions exceed the specified range?

jserv commented:
I would like to express a concern regarding the `rv_userdata` function being part of the public API. Originally, it was designed for internal implementation, which suggests there is no necessity to make `rv_userdata` accessible to users interacting with the public API. It might be more appropriate to keep this function internal to maintain clarity and simplicity in the API's usage.

ChinYikMing commented:
> Sure, a build macro can be added to `syscall_exit` and print the exit code only on test for passing CI.

The PRINT_EXIT_CODE build macro do not used in default build, thus make CI/host-x64 or CI/host-arm64 failed. Shall I add it to  default build?  Personally, I am prefer default build without PRINT_EXIT_CODE.

Another way to pass CI is that modifying the CI flow, build `rv32emu` with PRINT_EXIT_CODE  in CI tests.

jserv commented:
> > Sure, a build macro can be added to `syscall_exit` and print the exit code only on test for passing CI.
> 
> The PRINT_EXIT_CODE build macro do not used in default build, thus make CI/host-x64 or CI/host-arm64 failed. Shall I add it to default build? Personally, I am prefer default build without PRINT_EXIT_CODE.
> Another way to pass CI is that modifying the CI flow, build `rv32emu` with PRINT_EXIT_CODE in CI tests.

Drop macro `PRINT_EXIT_CODE` because it is not necessary to have multiple builds. Instead, let the shell or builtin commands to check exit code of RISC-V program emulated.

ChinYikMing commented:
> Due to the memory size being changed to a runtime-defined value, do we need to check if the addresses of load and store instructions exceed the specified range?

Sound reasonable. Maybe in future PR.

ChinYikMing commented on Makefile:
```diff
@@ -214,6 +210,7 @@ check: $(BIN)
 	)
 
 EXPECTED_aes_sha1 = 1242a6757c8aef23e50b5264f5941a2f4b4a347e  -
```
May I know the original input of this hash value?

ChinYikMing replied:
Maybe I guessed it

ChinYikMing commented:
> Drop macro `PRINT_EXIT_CODE` because it is not necessary to have multiple builds. Instead, let the shell or builtin commands to check exit code of RISC-V program emulated.

For minimal changes, printing `inferior exit code %d` inside `main` is able to pass the CI tests.

ChinYikMing commented:
> I would like to express a concern regarding the `rv_userdata` function being part of the public API. Originally, it was designed for internal implementation, which suggests there is no necessity to make `rv_userdata` accessible to users interacting with the public API. It might be more appropriate to keep this function internal to maintain clarity and simplicity in the API's usage.

Actually, we could introduce some helper functions or getter functions to access the specific field of userdata without losing the encapsulation. For example, some helper functions for memory viewing instead of returning whole userdata.

RinHizakura commented on src/main.c:
```diff
@@ -212,21 +210,33 @@ static void dump_test_signature(elf_t *elf)
+        .emu_data.vm_user = malloc(sizeof(vm_user_t)),
+        .cycle_per_step = 100,
+        .allow_misalign = opt_misaligned,
+    };
```
Since we are going to make these parameters user-defined now, I will suggest having some checks (`check_vm_attr()`) on the `vm_attr_t` before the emulator creation stage, otherwise bad parameters can lead to a crash which is not easy to figure out quickly. For example:  `mem_size` should be larger than `stack_size + args_offset_size`.

Note that `mem_size` should also be larger than the ELF mapping range to load correctly, but this may not be easy to check before the emulator creation stage. As an alternative, maybe applying the range check for `memory_write` at the emulator running stage is a simple idea to solve this. We may need to take care of the possible performance drop for the extra comparison on this solution.

ChinYikMing replied:
Nice perspective. I think validation of `vm_attr_t` could be complex. For example, shall we check the invalid file during validation or propagate to `elf_open` to verify the invalidation. Similar situation for memory, if the user-provided memory size is not enough to run emulation, shall we simply abort before `rv_create` or propagate to segmentation fault. We may need more discussion about this. Maybe in future PR?

RinHizakura replied:
Sure. We can leave these at other places.

ChinYikMing replied:
> Sure. We can leave these at other places.

I have mentioned it in the commit message.

RinHizakura commented on src/main.c:
```diff
@@ -246,33 +256,18 @@ int main(int argc, char **args)
     /* run based on the specified mode */
     if (opt_trace) {
-        run_and_trace(rv, elf);
+        run_and_trace(rv, opt_prog_name);
```
Why don't we just realize the elf name by `PRIV(rv)`? It seems that we don't need to pass `opt_prog_name` again here.

ChinYikMing replied:
Nice catch!

ChinYikMing replied:
As `PRIV(rv)` requires internal structure, `rv_userdata(rv)` as alternative.

RinHizakura commented on src/riscv.c:
```diff
@@ -105,11 +106,65 @@ riscv_word_t rv_get_reg(riscv_t *rv, uint32_t reg)
+void rv_remap_stdstream(riscv_t *rv, fd_stream_pair_t *fsp, uint32_t fsp_size)
+{
+    if (!rv)
+        return;
```
To make the implementation consistent, replace with `assert(rv)`.

RinHizakura commented on src/riscv.c:
```diff
@@ -105,11 +106,65 @@ riscv_word_t rv_get_reg(riscv_t *rv, uint32_t reg)
+
+    vm_attr_t *attr = PRIV(rv);
+    if (!attr)
+        return;
```
Similarly, replace with `assert(attr)`.

RinHizakura commented on src/riscv.c:
```diff
@@ -105,11 +106,65 @@ riscv_word_t rv_get_reg(riscv_t *rv, uint32_t reg)
+
+riscv_t *rv_create(const riscv_io_t *io, riscv_user_t userdata)
 {
     assert(io);
```
Let's `assert(attr)` here since we'll access it without checking in the future.

RinHizakura commented on src/riscv.c:
```diff
@@ -105,11 +106,65 @@ riscv_word_t rv_get_reg(riscv_t *rv, uint32_t reg)
+    for (uint32_t i = 0; i < fsp_size; i++) {
+        int fd = fsp[i].fd;
+        FILE *file = fsp[i].file;
+        if (fd != 0 && fd != 1 && fd != 2)
```
Maybe it's not a great idea to use 0, 1, and 2 here directly. @jserv Do you think we should define macros for each and expand them to the correct value considering the operating system(e.g.  `STDIN_FILENO` for POSIX and _fileno(stdin) on WIN)?

jserv replied:
> Do you think we should define macros for each and expand them to the correct value considering the operating system(e.g. `STDIN_FILENO` for POSIX and _fileno(stdin) on WIN)?

Agree. We shall proceed in a portable way.

ChinYikMing replied:
See [commit](https://github.com/sysprog21/rv32emu/pull/340/commits/f953ac1e051558623ae3d1bfd4cee70077e47c53#diff-f99b8d1f48ef2257a9bbbe6b487c621f74ef233c00858df5a37b06e89441e836)

jserv commented on src/riscv.c:
```diff
@@ -105,13 +114,63 @@ riscv_word_t rv_get_reg(riscv_t *rv, uint32_t reg)
+            continue;
+
+        /*
+         * check if stdio_fd exists or not
```
Refine the use of `stdio_fd`.

jserv commented on src/riscv.h:
```diff
@@ -97,6 +98,8 @@ enum {
 
 #define BLOCK_MAP_CAPACITY_BITS 10
 
+#define PRIV(x) ((vm_attr_t *) x->userdata)
```
The macro `PRIV` should not appear in public header, `riscv.h`.

ChinYikMing replied:
Stored it in `riscv_private.h`

jserv commented on src/riscv.h:
```diff
@@ -143,22 +146,19 @@ typedef struct {
 } riscv_io_t;
 
+/* run emulation */
+void run(riscv_t *rv);
```
To maintain a consistent API naming scheme and to prevent potential symbol conflicts, it is advisable to choose a name other than `run` for the function.

ChinYikMing replied:
How about `rv_start_emu`?

jserv replied:
> How about `rv_start_emu`?

Opt for concise and meaningful names. e.g., `rv_mainloop` and `rv_run`.

jserv replied:
Check the manner of GLib. The [GMainLoop struct](https://docs.gtk.org/glib/struct.MainLoop.html) is an opaque data type representing the main event loop of a GLib or GTK application.

jserv commented on src/riscv.h:
```diff
@@ -168,7 +168,19 @@ void rv_debug(riscv_t *rv);
 
-/* get RISC-V user data bound to an emulator */
+/*
+ * @rv: RICV-V
```
In this context, `rv` represents the instance of the RISC-V emulator.

ChinYikMing replied:
Since we drop the `rv_userdata`, no more comment about them.

jserv commented on src/riscv.h:
```diff
@@ -207,25 +227,82 @@ void rv_halt(riscv_t *rv);
+    char *elf_program;
+} vm_user_t;
+
+typedef struct {
```
Let's hold off on including the system emulation components until we are fully prepared to merge them.

jserv commented on src/riscv.h:
```diff
@@ -207,25 +227,82 @@ void rv_halt(riscv_t *rv);
+
+typedef union {
+    vm_user_t *vm_user;
+    vm_sys_t *vm_sys;
```
Ditto. Get rid of `vm_sys` and place proper TODO comments.

jserv commented on src/riscv.h:
```diff
@@ -207,25 +227,82 @@ void rv_halt(riscv_t *rv);
+     *
+     * use rv_remap_stdstream to overwrite them
+     */
+    int stdin_fd;
```
Unify the naming in the form of `fd_stdxxx`. That is,
```c
int fd_stdin, fd_stdout, fd_stderr;
```

jserv commented on src/riscv.h:
```diff
@@ -180,6 +192,14 @@ riscv_word_t rv_get_pc(riscv_t *rv);
+    FILE *file;
+} fd_stream_pair_t;
+
+/* remap stdio stream */
```
Refine the comment.

ChinYikMing commented:
> Actually, we could introduce some helper functions or getter functions to access the specific field of userdata without losing the encapsulation. For example, some helper functions for memory viewing instead of returning whole userdata.

Shall we drop it for now? Later on, we can think about what fields are gonna export to applications.

Note that the `run_and_trace` function requires program name to run for now, see [here](https://github.com/sysprog21/rv32emu/pull/340/files#diff-e0cf5b28d9b6b600f0af2bc78e8fd30ec675fd731a5da86f0c4283ffc0e40176). If we drop `rv_userdata`, the `run_and_trace` can only get program name by passing `opt_prog_name` or `attr` to it.

jserv commented:
> Note that the run_and_trace function requires program name to run for now, see [here](https://github.com/sysprog21/rv32emu/pull/340/files#diff-e0cf5b28d9b6b600f0af2bc78e8fd30ec675fd731a5da86f0c4283ffc0e40176). If we drop rv_userdata, the run_and_trace can only get program name by passing opt_prog_name or attr to it.

We could streamline the `run` and `run_and_trace` functions by treating `run_and_trace` as a specialized variant of `rv_run`, activated through specific callback functions or flags. This strategy would enable us to eliminate the need for exposing `rv_userdata` directly.

ChinYikMing commented:
> > Note that the run_and_trace function requires program name to run for now, see [here](https://github.com/sysprog21/rv32emu/pull/340/files#diff-e0cf5b28d9b6b600f0af2bc78e8fd30ec675fd731a5da86f0c4283ffc0e40176). If we drop rv_userdata, the run_and_trace can only get program name by passing opt_prog_name or attr to it.
> 
> We could streamline the `run` and `run_and_trace` functions by treating `run_and_trace` as a specialized variant of `rv_run`, activated through specific callback functions or flags. This strategy would enable us to eliminate the need for exposing `rv_userdata` directly.

Since we have introduced `vm_attr_t`, we can simply add one more field (e.g., `trace`). `rv_run` can check that easily.

ChinYikMing commented:
> Since we have introduced `vm_attr_t`, we can simply add one more field (e.g., `trace`). `rv_run` can check that easily.

Another way: we can introduce `rv_set_main_loop(rv, void callback(void *arg), void *arg)`. The application can pass their own callback to register main loop. In this way, we can simply call their callback and decouple for many scenario. Inspired by [emscripten](https://emscripten.org/docs/api_reference/emscripten.h.html?highlight=main_loop#c.emscripten_set_main_loop). If callback is NULL, we can fallback to normal `rv_run`. If callback is NULL and `trace` field in `vm_attr_t` is set, then fallback to `rv_run_and_trace`. Specifically, `rv_run` and `rv_run_and_trace` are private function in "riscv.c".

jserv commented:
> Since we have introduced `vm_attr_t`, we can simply add one more field (e.g., `trace`). `rv_run` can check that easily.

Exactly. It is time to drop `rv_userdata` by extensive use of `vm_attr_t`. We may refine the term "userdata" as well to avoid unintended confusion.

jserv commented:
> Another way: we can introduce `rv_set_main_loop(void callback(void *arg), void *arg)`. The application can pass their own callback to register main loop. In this way, we can simply call their callback and decouple for many scenario. Inspired by [emscripten](https://emscripten.org/docs/api_reference/emscripten.h.html?highlight=main_loop#c.emscripten_set_main_loop). If callback is NULL, we can fallback to normal `rv_run`. If callback is NULL and `trace` field in `vm_attr_t` is set, then fallback to `rv_run_and_trace`. Specifically, `rv_run` and `rv_run_and_trace` are private function in "riscv.c".

Including the callback function as part of the emulator's context is possible, but it’s not necessary to have a separate function like `rv_set_main_loop` for assigning callbacks. Specifying the callback function directly using C99 designated initializers is more efficient, especially since it is unlikely we'll need to change the callback during the execution of the RISC-V emulation main loop.

jserv commented on src/riscv.h:
```diff
@@ -207,27 +210,97 @@ void rv_halt(riscv_t *rv);
+    int error;
+
+    /* TODO: for logging feature */
+    int logging_level;
```
Rename to `log_level` for short.

jserv commented on src/riscv_private.h:
```diff
@@ -100,7 +102,7 @@ struct riscv_internal {
 
     /* user provided data */
-    riscv_user_t userdata;
+    riscv_user_t attr;
```
Would it be better to name this field as `priv` or `data`?

ChinYikMing replied:
I think `data` is decent.

jserv commented on src/riscv.h:
```diff
@@ -207,27 +210,97 @@ void rv_halt(riscv_t *rv);
+typedef union {
+    vm_user_t *vm_user;
+    /* TODO: system emulator stuff */
+} vm_emu_data_t;
```
It is quite confusing to have `vm` and `emu` here.

ChinYikMing replied:
How about `vm_data_t`? The instance of it can simply called `data` in `vm_attr_t` struct. Hopefully, it will not confuse with the `data` in internal implementation. I think it is not because the latter is private.

jserv replied:
> How about `vm_data_t`? 

It looks good to me.

RinHizakura commented on src/riscv.c:
```diff
@@ -105,24 +108,112 @@ riscv_word_t rv_get_reg(riscv_t *rv, uint32_t reg)
+        int new_fd = fileno(file);
+#else
+        int new_fd = _fileno(file);
+#endif
```
How about adding another macro like `FILENO()` here for both, so we can make the codes here clean? Then you can define platform-specific `FILENO()`  at the place you defined things like `STDIN_FILENO`.

ChinYikMing replied:
Got it.

ChinYikMing replied:
Check this [commit](https://github.com/sysprog21/rv32emu/pull/340/commits/ea51b2d1300c3d2379c4274af1faa74423fece89#diff-f99b8d1f48ef2257a9bbbe6b487c621f74ef233c00858df5a37b06e89441e836) out.

RinHizakura reviewed (approved):
I'm good with the changes now! Let's see if other guys have more comments.

Thank you!

jserv reviewed (changes requested):
In your git commit messages, it is more appropriate to mention "this commit" rather than "this PR," as "PR" refers to a pull request, not a commit.

ChinYikMing committed:
Refine the API in the public header

The following should be included in an emulator's simple and clear
public API:
1. create/init core
2. run emulation
3. delete/destroy core

Other components, including as memory, file systems, program data,
etc., should be abstracted from the user, as a result, setting a
configuration value (vm_attr_t) is sufficient. The user needs to
manage about memory (state_t) and elf stuff before this commit.
The user can just construct a core, run it, and shut it down
after this commit, so they won't need to worry about them anymore.

The vm_attr_t has multiple fields and they are commented clearly
in the code. As you can see in "main", there are various mode to run the
emulator such as "run_and_trace", "gdbstub", and "profiling". Thus,
a field call "run_flag" is introduced in vm_attr_t.

For standard stream remapping, rv_remap_stdstream function is
introduced. The emulator can remap default standard stream to required
streams after creating the emulator by calling the rv_remap_stdstream
function.

rv_userdata has been dropped since PRIV macro is sufficient for
internal implemntation. Also, application will not need to access it
directly.

elf is reopened in dump_test_signature because elf is allocated during
rv_create. It is acceptable to reopen elf since it is only for testing.

Print inferior exit code to console inside main instead of syscall_exit
because the actual usage of exit code depends on applications of using
riscv public API.

The io interface is not changed in this commit because it could maybe
reused with semu in some way, still need to be investigated. Logging
feature and system emulator integration are not implemented yet. Also,
a validator for validating the user-defined vm_attr_t might need to be
introduced.

related: #310

visitorckw commented:
The changes also look good to me. Thanks!
